### PR TITLE
Add cluster and np templating support for Azure

### PR DESCRIPTION
**Pull request by axbarsan**
Closes https://github.com/giantswarm/giantswarm/issues/13243

This PR adds support for Azure V5 for the `template cluster` and `template nodepool` commands.

There are a lot of changes because the existing AWS implementation had to be refactored to support multiple providers. The AWS logic is untouched.

I used release `v12.1.0` in the examples below, because that's the latest active release in the releases repo, in production only a release version newer than `v12.2.0` is valid.

The CRs produced:

<details>
<summary>Azure cluster template output</summary>

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
kind: AzureCluster
metadata:
  creationTimestamp: null
  labels:
    azure-operator.giantswarm.io/version: 4.2.0
    cluster.x-k8s.io/cluster-name: i33b3
    giantswarm.io/cluster: i33b3
    giantswarm.io/organization: giantswarm
    release.giantswarm.io/version: 12.1.0
  name: i33b3
  namespace: default
spec:
  controlPlaneEndpoint:
    host: api.i33b3.k8s.ghost.westeurope.azure.gigantic.io
    port: 443
  location: westeurope
  networkSpec:
    vnet:
      name: ""
status:
  bastion:
    image: {}
    osDisk:
      diskSizeGB: 0
      managedDisk:
        storageAccountType: ""
      osType: ""
  network:
    apiServerIp: {}
    apiServerLb:
      backendPool: {}
      frontendIpConfig: {}
  ready: false
---
apiVersion: cluster.x-k8s.io/v1alpha3
kind: Cluster
metadata:
  annotations:
    cluster.giantswarm.io/description: Adrian test
  creationTimestamp: null
  labels:
    azure-operator.giantswarm.io/version: 4.2.0
    cluster-operator.giantswarm.io/version: 0.23.16
    cluster.x-k8s.io/cluster-name: i33b3
    giantswarm.io/cluster: i33b3
    giantswarm.io/organization: giantswarm
    release.giantswarm.io/version: 12.1.0
  name: i33b3
  namespace: default
spec:
  clusterNetwork:
    apiServerPort: 443
    serviceDomain: i33b3.k8s.ghost.westeurope.azure.gigantic.io
    services:
      cidrBlocks:
      - 172.31.0.0/16
  controlPlaneEndpoint:
    host: api.i33b3.k8s.ghost.westeurope.azure.gigantic.io
    port: 443
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
    kind: AzureCluster
    name: i33b3
    namespace: default
status:
  controlPlaneInitialized: false
  infrastructureReady: false
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
kind: AzureMachine
metadata:
  creationTimestamp: null
  labels:
    azure-operator.giantswarm.io/version: 4.2.0
    cluster.x-k8s.io/cluster-name: i33b3
    cluster.x-k8s.io/control-plane: "true"
    giantswarm.io/cluster: i33b3
    giantswarm.io/organization: giantswarm
    release.giantswarm.io/version: 12.1.0
  name: i33b3-master-0
  namespace: default
spec:
  availabilityZone: {}
  failureDomain: "1"
  image:
    marketplace:
      offer: flatcar-container-linux-free
      publisher: kinvolk
      sku: stable
      version: 2345.3.1
  location: westeurope
  osDisk:
    diskSizeGB: 50
    managedDisk:
      storageAccountType: Premium_LRS
    osType: Linux
  sshPublicKey: <REDACTED>
  vmSize: Standard_D4_v3
status:
  ready: false
```
</details>

<details>
<summary>Azure nodepool template output</summary>

```yaml
apiVersion: exp.infrastructure.cluster.x-k8s.io/v1alpha3
kind: AzureMachinePool
metadata:
  creationTimestamp: null
  labels:
    azure-operator.giantswarm.io/version: 4.2.0
    cluster.x-k8s.io/cluster-name: i33b3
    giantswarm.io/cluster: i33b3
    giantswarm.io/machine-pool: g6x9q
    giantswarm.io/organization: giantswarm
    release.giantswarm.io/version: 12.1.0
  name: g6x9q
spec:
  location: westeurope
  template:
    osDisk:
      diskSizeGB: 0
      managedDisk:
        storageAccountType: ""
      osType: ""
    sshPublicKey: <REDACTED>
    vmSize: Standard_D4_v3
status:
  ready: false
  replicas: 0
---
apiVersion: exp.infrastructure.cluster.x-k8s.io/v1alpha3
kind: MachinePool
metadata:
  annotations:
    cluster.k8s.io/cluster-api-autoscaler-node-group-max-size: "10"
    cluster.k8s.io/cluster-api-autoscaler-node-group-min-size: "3"
    machine-pool.giantswarm.io/name: Adrian test
  creationTimestamp: null
  labels:
    azure-operator.giantswarm.io/version: 4.2.0
    cluster-operator.giantswarm.io/version: 0.23.16
    cluster.x-k8s.io/cluster-name: i33b3
    giantswarm.io/cluster: i33b3
    giantswarm.io/machine-pool: g6x9q
    giantswarm.io/organization: giantswarm
    release.giantswarm.io/version: 12.1.0
  name: g6x9q
spec:
  clusterName: i33b3
  failureDomains:
  - "1"
  replicas: 10
  template:
    metadata: {}
    spec:
      bootstrap: {}
      clusterName: i33b3
      infrastructureRef:
        apiVersion: exp.infrastructure.cluster.x-k8s.io/v1alpha3
        kind: AzureMachinePool
        name: g6x9q
status:
  bootstrapReady: false
  infrastructureReady: false
  replicas: 0
---
apiVersion: core.giantswarm.io/v1alpha1
kind: Spark
metadata:
  creationTimestamp: null
  labels:
    cluster.x-k8s.io/cluster-name: i33b3
    giantswarm.io/cluster: i33b3
    release.giantswarm.io/version: 12.1.0
  name: g6x9q
spec: {}
status:
  dataSecretName: ""
  failureMessage: ""
  failureReason: ""
  ready: false
  verification:
    algorithm: ""
    hash: ""
```
</details>

### CODE REVIEW

File: cmd/template/cluster/aws.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/giantswarm/kubectl-gs/pkg/clusterlabels"
+	"github.com/giantswarm/kubectl-gs/pkg/release"
+	"github.com/giantswarm/microerror"
+	"io"
+	"sigs.k8s.io/yaml"
+	"strings"
+	"text/template"
+)
+
+func writeAWSTemplate(out io.Writer, name string, flags *flag) error {
```
Comment (axbarsan) [inline cmd/template/cluster/aws.go:15]: Untouched previous implementation

File: cmd/template/cluster/flag.go [hunk 2]
```diff
@@ -94,45 +126,88 @@ func (f *flag) Validate() error {
...
-		if !aws.ValidateAZ(f.Region, az) {
-			return microerror.Maskf(invalidFlagError, "The AZ name %q passed via --%s is not a valid AZ name for region %s", az, flagMasterAZ, f.Region)
+	{
+		if f.Provider == key.ProviderAzure {
+			if len(f.AzurePublicSSHKey) < 1 {
+				return microerror.Maskf(invalidFlagError, "--%s must not be empty on Azure", flagAzurePublicSSHKey)
+			} else {
+				sshKey := []byte(f.AzurePublicSSHKey)
+				dest := make([]byte, base64.StdEncoding.EncodedLen(len(sshKey)))
+				_, err = base64.StdEncoding.Decode(dest, sshKey)
```
Comment (tuommaki) [inline cmd/template/cluster/flag.go:155]: There's `EncodedLen` w/ `Decode` here - is that intentional?

File: cmd/template/nodepool/flag.go [hunk 3]
```diff
@@ -86,62 +131,127 @@ func (f *flag) Validate() error {
...
-		if len(azs) < 1 {
-			return microerror.Maskf(invalidFlagError, "--%s must be configured with at least 1 AZ", flagAvailabilityZones)
+	{
+		if f.Provider == key.ProviderAzure {
+			if len(f.AzurePublicSSHKey) < 1 {
+				return microerror.Maskf(invalidFlagError, "--%s must not be empty on Azure", flagAzurePublicSSHKey)
+			} else {
+				sshKey := []byte(f.AzurePublicSSHKey)
+				dest := make([]byte, base64.StdEncoding.EncodedLen(len(sshKey)))
+				_, err = base64.StdEncoding.Decode(dest, sshKey)
```
Comment (tuommaki) [inline cmd/template/nodepool/flag.go:160]: Same as earlier, is `EncodedLen` w/ `Decode` correct?

**Review by tuommaki — APPROVED** [review]
Skimmed through, looks good. Few points might have some weirdness with base64 length.

File: cmd/template/cluster/flag.go (hunk 2 above)
Comment (axbarsan) [reply cmd/template/cluster/flag.go]: You're right, it's supposed to be `DecodedLen` 😅 

But turns out there's `DecodeString()` which does exactly the same thing I wanted to do, so I'll switch to that
https://golang.org/src/encoding/base64/base64.go?s=9757:9816#L373

**Pull request merged**